### PR TITLE
[unittests] Create OperationGradTest to isolate unit tests for gradient computation

### PR DESCRIPTION
**Pull request by d1jang**
*Description*:
Please see the comment for computeGradient().

*Testing*:
Run GradCheckTest

*Documentation*:

Partially fixing #2513

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: tests/unittests/GradCheckTest.cpp [hunk 1]
```diff
@@ -168,36 +168,135 @@ void performGradCheck(ExecutionEngine &EE, PlaceholderBindings &bindings,
...
+
+  auto *A =
+      mod.createPlaceholder(ElemKind::FloatTy, {1, numOutputElem}, "A", false);
+  bindings.allocate(A)->getHandle() = {1, 2};
+  auto *B =
+      mod.createPlaceholder(ElemKind::FloatTy, {1, numOutputElem}, "B", false);
+  bindings.allocate(B)->getHandle() = {3, 4};
+
+  Node *O = F->createAdd("add", A, B);
+  // O = { { a1 + b1 }, { a2 + b2 } }
```
Comment (artemrakhov) [inline tests/unittests/GradCheckTest.cpp]: We have similar but different comment in the actual grad computing code:
https://github.com/pytorch/glow/blob/2793672d9368f47690208acc3c8933944136cef9/lib/Optimizer/Lower.cpp#L47

File: tests/unittests/GradCheckTest.cpp [hunk 2]
```diff
@@ -168,36 +168,135 @@ void performGradCheck(ExecutionEngine &EE, PlaceholderBindings &bindings,
   EXPECT_FALSE(allZeroGradients);
 }
 
-TEST_P(InterpreterGrad, gradientCheckConcat) {
+/// Compute gradient of inputVar based on the one of outputNode.
```
[the remark this answers is not in the document]
Comment (d1jang) [reply tests/unittests/GradCheckTest.cpp]: I updated the comment on my best understanding. Please let me know if the new comment looks less confusing.

File: tests/unittests/GradCheckTest.cpp (hunk 1 above)
Comment (d1jang) [reply tests/unittests/GradCheckTest.cpp]: Thanks! I'll look into that and let me know if the style of this comment should be unified.

**Comment by d1jang** [comment]
I addressed the issues you mentioned. Created a new test suite, OperationGradTest, and created a unittest for FC. Please take another look at the most recent commit.

File: tests/unittests/OperatorGradTest.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  // W = [[w1_1,w1_2],[w2_1,w2_2]]
+  // O = [O1, O2] = [x1*w1_1 + x2*w2_1, x1*w2_1 + x2*w2_2] = [8, 11]
+  // dO_k/dWi_j = (k == j ? 1 : 0) * x_i
+  // df/dW_i_j = sum_of_all_k (dF/dO_k * dO_k/dW_i_j).
+  // dF/dO = [ 8, 11 ]
+  // dF/dW = [ [O1*x1, O2*x1], [O1*x2, O2*x2] ]
+  // dF/dW = [ [8*1,   11*1 ], [8*2,   11*2 ] ]
+  expected.getHandle() = {8 * 1, 11 * 1, 8 * 2, 11 * 2};
+  EXPECT_TRUE(expected.isEqual(*getGradTensor(varGrads, W)));
+}
```
Comment (artemrakhov) [inline tests/unittests/OperatorGradTest.cpp]: Please add a TODO for grads of x and b

File: tests/unittests/OperatorGradTest.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  /// Compute gradients of placeholders in bindings_. Given
+  /// outputNode, representing F(Vars) = q, this function creates an L2 node
+  /// computing its MSE against zeros, G(Vars) = 0.5 * /F(Vars)^2 and computes
+  /// dG/dVar with respect to all variables in "main" function.
+  ///
+  /// Note that dG/dF = F(Vars), the same as outputNode's forward value.
+  ///
+  /// \param outputNode Node that contains result of f(x).
+  VariableGradientsList computeVarGrads(Node *outputNode) {
+    // A RegressionNode is created just to generate a fake gradient with the
```
Comment (artemrakhov) [inline tests/unittests/OperatorGradTest.cpp]: I suggest to remove this comment, because we already explain purpose of RegressionNode in comment above.

File: tests/unittests/OperatorGradTest.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    // Create a version of the network that records the gradients to some side
+    // table instead of updating them.
+    VariableGradientsList varGrads;
+    TrainingConfig TC;
+    Function *recordNet = glow::differentiate(F_, TC, "record", &varGrads);
+    allocateGrads(varGrads);
+    EE_.compile(CompilationMode::Train, recordNet);
+
+    // Train the network just once to record the values of gradient for
+    // inputVar.
```
Comment (artemrakhov) [inline tests/unittests/OperatorGradTest.cpp]: inputVar -> all input variables.

File: tests/unittests/OperatorGradTest.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+ * limitations under the License.
+ */
+
+#include "BackendTestUtils.h"
+
+#include "glow/Base/Tensor.h"
+#include "glow/ExecutionEngine/ExecutionEngine.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/Nodes.h"
+#include "glow/IR/IR.h"
```
Comment (artemrakhov) [inline tests/unittests/OperatorGradTest.cpp]: Why do you need to include IR/ ?

File: tests/unittests/OperatorGradTest.cpp (hunk 3 above)
Comment (d1jang) [reply tests/unittests/OperatorGradTest.cpp]: Done.

File: tests/unittests/OperatorGradTest.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+#include "gtest/gtest.h"
+
+using namespace glow;
+
+class OperatorGradTest : public BackendTest {
+protected:
+  /// Compute gradients of placeholders in bindings_. Given
+  /// outputNode, representing H(Vars), this function will train H(vars) to be
+  /// 0. It is achieved by creating RegressionNode between outputNode and 0,
+  /// which will minimize F, a total squared error of H devided by 2.
```
Comment (artemrakhov) [inline tests/unittests/OperatorGradTest.cpp]: devided -> divided

Comment (d1jang) [reply tests/unittests/OperatorGradTest.cpp]: Done.
